src/daemon/auth/providers/openid_connect/provider.rs

```diff
@@ -0,0 +1,1123 @@
+        // what we just learned and using the credentials we read from config
+        // above.
+        let client = FlexibleClient::from_provider_metadata(meta, client_id, Some(client_secret));
+
+        // Note: we still haven't actually verified that the client id and
+        // secret are correct, that will only happen when we try to exchange a
+        // temporary code for access and id tokens.
+
+        // Configure the client to instruct the 3rd party login form that after
+        // successful login it should redirect, via the client browser, to the
+        // Krill authentication callback endpoint. When the callback is invoked
+        // it will come through to us so that we can exchange the temporary code
+        // for access and id tokens.
+        let redirect_uri = RedirectUrl::new(
+            self.config
+                .service_uri()
+                .join(AUTH_CALLBACK_ENDPOINT.trim_start_matches('/').as_bytes())
+                .to_string(),
+        )?;
+
+        // Log the redirect URI to help the operator in the event that the
+        // OpenID Connect: provider complains that the redirect URI doesn't match
+        // that configured at the provider.
+        debug!("OpenID Connect: Redirect URI set to {}", redirect_uri.to_string());
+
+        let client = client.set_redirect_uri(redirect_uri);
+
+        Ok(client)
+    }
+
+    /// Build a logout URL to which the client should be directed to so that
+    /// they can logout with the OpenID Connect: provider. The URL includes an
+    /// OpenID Connect: RP Initiatiated Logout spec compliant query parameter
+    /// telling the provider to redirect back to Krill once logout is complete.
+    ///
+    /// See: https://openid.net/specs/openid-connect-rpinitiated-1_0.html
+    fn build_logout_url(&self, meta: &WantedMeta) -> KrillResult<String> {
+        let service_uri = self.config.service_uri();
+        let logout_url = if let Some(url) = &meta.additional_metadata().end_session_endpoint {
+            format!("{}?post_logout_redirect_uri={}", url, service_uri.as_str())
+        } else if meta.additional_metadata().revocation_endpoint.is_some() {
+            service_uri.to_string()
+        } else if let Some(url) = &self.oidc_conf()?.logout_url {
+            url.to_string()
+        } else {
+            // should be unreachable due to checks done in discover().
+            unreachable!()
+        };
+
+        debug!("OpenID Connect: Logout URL will be {:?}", &logout_url);
+
+        Ok(logout_url)
+    }
+
+    fn try_refresh_token(&self, session: &ClientSession) -> Option<Auth> {
+        match &session.secrets.get(0) {
+            Some(refresh_token) => {
+                debug!("OpenID Connect: Refreshing token for user: \"{}\"", &session.id);
+                trace!("OpenID Connect: Submitting RFC-6749 section 6 Access Token Refresh request");
+                match self.conn.read() {
```

Comment:
**TODO:** Review this locking code, it feels clumsy. Is the Arc needed given that the provider is accessed through Request::state which is itself `Arc<RwLock<KrillServer>>` ?